### PR TITLE
HTML API: Return subclass from ::create_fragment

### PR DESCRIPTION
**Pull request by sirreal**
Trac ticket: Core-61374

In a subclass of `WP_HTML_Processor` when an instance is created from the static `::create_fragment` method the instance will be an instance of `WP_HTML_Processor` and _not_ and instance of the subclass:

```php
class Sub extends WP_HTML_Processor {}
$processor = Sub::create_fragment( '<div>' );
get_class( $processor ); // "WP_HTML_Processor" !!
```

It's very unlikely that the parent class instance is intended to be created from the subclass.

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: tests/phpunit/tests/html-api/wpHtmlProcessorSubclass.php [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+	 * @ticket TBD
+	 */
+	public function test_subclass_create_fragment_creates_subclass() {
+		$processor = WP_HTML_Processor::create_fragment( '' );
+		$this->assertInstanceOf( WP_HTML_Processor::class, $processor, '::create_fragment did not return class instance.' );
+
+		$processor = WP_HTML_Processor_Subclass::create_fragment( '' );
+		$this->assertInstanceOf( WP_HTML_Processor_Subclass::class, $processor, '::create_fragment did not return subclass instance.' );
+	}
+}
```
Comment (dmsnell) [inline tests/phpunit/tests/html-api/wpHtmlProcessorSubclass.php]: maybe we should put this into the `wpHtmlProcessor.php` test file and create an anonymous subclass? I  don't think it deserves its own file just for this one test.

Comment (sirreal) [reply tests/phpunit/tests/html-api/wpHtmlProcessorSubclass.php]: Agreed and done.

File: tests/phpunit/tests/html-api/wpHtmlProcessor.php [hunk 2]
```diff
@@ -517,4 +517,23 @@ public static function data_html_with_target_element_and_depth_of_next_node_in_b
 			'Funky comment'                     => array( '<div><p>What <br class="target"><//wp:post-author></p></div>', 5 ),
 		);
 	}
+
+	/**
+	 * Ensures that subclasses can be created from ::create_fragment method.
+	 *
+	 * @ticket TBD
```
Comment (sirreal) [inline tests/phpunit/tests/html-api/wpHtmlProcessor.php]:
```suggestion
	 * @ticket 61374
```

File: tests/phpunit/tests/html-api/wpHtmlProcessor.php [hunk 3]
```diff
@@ -517,4 +517,23 @@ public static function data_html_with_target_element_and_depth_of_next_node_in_b
...
+		$processor = WP_HTML_Processor::create_fragment( '' );
+		$this->assertInstanceOf( WP_HTML_Processor::class, $processor, '::create_fragment did not return class instance.' );
+
+		$subclass_instance = new class('') extends WP_HTML_Processor {
+			public function __construct( $html ) {
+				parent::__construct( $html, parent::CONSTRUCTOR_UNLOCK_CODE );
+			}
+		};
+
+		$subclass_processor = call_user_func( array( get_class( $subclass_instance ), 'create_fragment' ), '' );
```
Comment (dmsnell) [inline tests/phpunit/tests/html-api/wpHtmlProcessor.php:536]: interesting! I didn't think of the way we would have to bend to call the static methods. nice work. I did learn before that PHP creates new actual classes with auto-generated names in them for the anonymous classes, so tis probably maps to some string value (speaking of `get_class()`)

**Review by dmsnell — APPROVED** [review]

**Comment by dmsnell** [comment]
Merged in [[58363]](https://core.trac.wordpress.org/changeset/58363)
https://github.com/wordpress/wordpress-develop/commit/0cad031a5afb33567b219ffcaeb62dfff6bd0936

**Pull request closed without merging**